packages/babel-helper-module-transforms/src/rewrite-live-references.js

```diff
@@ -306,4 +306,41 @@ const rewriteReferencesVisitor = {
       }
     },
   },
+  "ForOfStatement|ForInStatement"(path) {
+    const { scope, node } = path;
+    const { left } = node;
+    const { exported, scope: programScope } = this;
+
+    // if it's a variable declaration, such as for (let {foo} of []) {}
+    // then no transformation is needed
+    if (!t.isVariableDeclaration(left)) {
+      let didTransform = false;
+      const bodyPath = path.get("body");
+      const loopBodyScope = bodyPath.scope;
+      for (const name of Object.keys(t.getOuterBindingIdentifiers(left))) {
+        if (
+          exported.get(name) &&
+          programScope.getBinding(name) === scope.getBinding(name)
+        ) {
+          didTransform = true;
+          if (loopBodyScope.hasOwnBinding(name)) {
+            loopBodyScope.rename(name);
+          }
+        }
+      }
+      if (!didTransform) {
+        return;
+      }
+      const newLoopId = t.identifier(scope.generateUidBasedOnNode(left));
```

Comment:
```suggestion
      const newLoopId = scope.generateUidIdentifierBasedOnNode(left);
      scope.push({ id: newLoopId });
```

The first change is for shorter code size. The second change is to make sure the scope is synced with AST changes.

Reply:
@JLHwung The second change seems to lead to broken output? For example:
```
for (foo of []) {}
```
is transformed into
```
for (let _foo of []) {
  exports.bar = exports.foo = foo = _foo;

  var _foo;
}
```
after the second change. I don't think the `var _foo;` should be there?

Reply:
Yeah, `push()` declares a new variable. We need `registerBinding()` or `registerDeclaration()`.